### ISSUE TITLE
Second run false-positive with anonymous class (caching issue)

### ISSUE THREAD

**Issue by weirdan** [vimeo/psalm#532]
```php
<?php
namespace Weirdan;

interface I {}

class C
{
    public function getI(): I
    {
        return new class implements I 
        { 
        
        };
    }
}
```
Running psalm twice produces the following output:
```
~/src/t❯ psalm --clear-cache
Cache directory deleted
~/src/t❯ psalm
Scanning files...
Analyzing files...
Checks took 0.05 seconds and used 16.395MB of memory
Psalm was able to infer types for 100.000% of the codebase
~/src/t❯ psalm
Scanning files...
Analyzing files...
ERROR: InvalidReturnType - src/C.php:8:29 - The declared return type 'Weirdan\I' for Weirdan\C::getI is incorrect, got '_home_weirdan_src_t_src_C_php_10'
    public function getI(): I
    {
        return new class implements I
        {

        };
    }


ERROR: InvalidReturnStatement - src/C.php:10:9 - The type '_home_weirdan_src_t_src_C_php_10' does not match the declared return type 'Weirdan\I' for Weirdan\C::getI
        return new class implements I
        {

        };


Checks took 0.05 seconds and used 14.299MB of memory
Psalm was able to infer types for 100.000% of the codebase
~/src/t❯
```
Where `_home_weirdan_src_t_src_C_php_10` is the source line where anonymous class is defined.

Running any number of times with `--no-cache` results (properly) in no issues.

**Comment by muglug** [comment]
Thanks for ticketing! I think I've fixed. I need to add tests that properly mock out the effects of caching this storage.

**Comment by weirdan** [comment]
Seems to be working fine now, thanks!
